cairo/src/contracts/token/extensions/hyp_erc721_URI_storage.cairo

```diff
@@ -1,48 +1,213 @@
-        fn initialize(ref self: ContractState) {}
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        #[flat]
+        OwnableEvent: OwnableComponent::Event,
+        #[flat]
+        MailboxclientEvent: MailboxclientComponent::Event,
+        #[flat]
+        RouterEvent: RouterComponent::Event,
+        #[flat]
+        GasRouterEvent: GasRouterComponent::Event,
+        #[flat]
+        TokenRouterEvent: TokenRouterComponent::Event,
+        #[flat]
+        ERC721Event: ERC721Component::Event,
+        #[flat]
+        SRC5Event: SRC5Component::Event,
+        #[flat]
+        HypErc721Event: HypErc721Component::Event,
+        #[flat]
+        UpgradeableEvent: UpgradeableComponent::Event,
+    }
 
-        fn balance_of(self: @ContractState, account: u256) -> u256 {
-            0
-        }
+    #[constructor]
+    fn constructor(
+        ref self: ContractState,
+        owner: ContractAddress,
+        mailbox: ContractAddress
+    ) {
+        self.ownable._transfer_ownership(owner);
+        self.mailboxclient.initialize(mailbox, Option::None, Option::None);
+    }
 
-        fn token_uri(self: @ContractState, token_id: u256) -> u256 {
-            0
+    #[abi(embed_v0)]
+    impl HypERC721URIStorageImpl of super::IHypERC721URIStorage<ContractState> {
+        fn initialize(
+            ref self: ContractState,
+            _mint_amount: u256,
+            _name: ByteArray,
+            _symbol: ByteArray,
+            _hook: ContractAddress,
+            _interchainSecurityModule: ContractAddress,
+            owner: ContractAddress
+        ) {
+            self.ownable.initializer(owner);
+            let mailbox = self.mailboxclient.mailbox.read();
+            self.mailboxclient.initialize(mailbox.contract_address, Option::Some(_hook), Option::Some(_interchainSecurityModule));
+            self.hyp_erc721.initialize(
+                _mint_amount,
+                _name,
+                _symbol
+            );
         }
+    }
 
-        fn supports_interface(self: @ContractState, interface_id: u256) -> bool {
-            false
+    #[abi(embed_v0)]
+    impl HypErc721Upgradeable of IUpgradeable<ContractState> {
+        fn upgrade(ref self: ContractState, new_class_hash: starknet::ClassHash) {
+            self.ownable.assert_only_owner();
+            self.upgradeable.upgrade(new_class_hash);
         }
     }
 
     #[generate_trait]
     impl InternalImpl of InternalTrait {
         fn transfer_from_sender(ref self: ContractState, token_id: u256) -> u256 {
-            0
+            self.hyp_erc721.transfer_from_sender(token_id);
+            token_id
         }
 
-        fn transfer_to(ref self: ContractState, recipient: u256, token_id: u256, token_uri: u256) {}
+        fn transfer_to(ref self: ContractState, recipient: ContractAddress, token_id: u256, token_uri: u256) {
+            self.hyp_erc721.transfer_to(recipient, token_id);
+        }
 
         fn before_token_transfer(
```

Comment:
This needs to be overriding [this](https://github.com/OpenZeppelin/cairo-contracts/blob/24aacb42b8ae102cdf4af58a1376d71d4eef974f/packages/token/src/erc721/erc721.cairo#L93) interface as we override the `TokenRouterHooksTrait` below

Reply:
Actually solidity one seems to only be choosing the impl of `ERC721EnumerableUpgradeable`. I guess we also need to make our hyp_erc721 to use [erc721_enumerable](https://github.com/OpenZeppelin/cairo-contracts/blob/main/packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo). What do you think? @ametel01

Reply:
Yes we should actually but since we are on OZ 0.14 and it is not available there we need to implement it.

I'm not working on any issues at the moment so i'll work on this